tensorly/decomposition/CMTF_ALS

```diff
@@ -0,0 +1,165 @@
+    order = tl.argmax(perm, axis=1)
+    perm = perm[:, order]  # perm[r,r] now corresponds to \prod_i (a^i_r.T @ ahat^i_r) / (||(a^i_r|| ||ahat^i_r||)
+    V_hat = V_hat[:, order]
+    xi_hat = xi_hat[order]
+
+    # calculate factor match score
+    FMS = tl.zeros(R)
+    for r in range(R):
+        FMS[r] = (1 - tl.abs(xi[r] - xi_hat[r]) / tl.max([xi[r], xi_hat[r]])) * tl.abs(
+            perm[r, r] * tl.dot(V[:, r], V_hat[:, r])) / (tl.norm(V[:, r]) * tl.norm(V_hat[:, r]))
+    return tl.min(FMS)
+
+
+def cmtf_als_for_third_order_tensor(X, Y, R):
+    """
+    Calculates a coupled matrix and tensor factorization of 3rd order tensor and matrix which are coupled in first mode.
+
+    Assume you have 3rd order tensor X = [[A,B,C]] and matrix Y = A @ V.T, which are coupled in 1st mode.
+    With coupled matrix and tensor factorization (CTMF), the factor matrices A, B, C for the CP decomposition of X
+    and the matrix V are found.
+    This implementation only works for a coupling in the first mode.
+
+    Solution is found via alternating least squares (ALS) as described in Figure 5 of
+    @article{acar2011all,
+      title={All-at-once optimization for coupled matrix and tensor factorizations},
+      author={Acar, Evrim and Kolda, Tamara G and Dunlavy, Daniel M},
+      journal={arXiv preprint arXiv:1105.3422},
+      year={2011}
+    }
+
+    Parameters
+    ----------
+    X : tl.tensor
+        3rd order tensor with factor matrices [[A, B, C]]
+    Y : tl.tensor
+        matrix that is coupled with tensor in first mode: Y = A @ V.T
+    R : int
+        rank for CP decomposition of X
+
+    Returns
+    -------
+    A, B, C, V : estimated matrices
+
+    Examples
+    --------
+    A = tl.tensor([[1, 2], [3, 4]])
+    B = tl.tensor([[1, 0], [0, 2]])
+    C = tl.tensor([[2, 0], [0, 1]])
+    V = tl.tensor([[2, 0], [0, 1]])
+    R = 2
+
+    X = tl.kruskal_to_tensor([None, [A, B, C]])  # weights are None
+    Y = A @ V.T
+
+    A_hat, B_hat, C_hat, V_hat = cmtf_als_for_third_order_tensor(X, Y, R)
+
+    """
+
+    # initialize values
+    shape_list = X.shape
+    s = shape_list + (Y.shape[1],)
+    A, B, C, V = tl.random.random_kruskal(s, R).factors
+    f_old = np.inf
+
+    # alternating least squares
+    # note that no rescaling is done since it is not guaranteed that the columns in true matrices have unit norm
+    # note that the order of the khatri rao product is reversed since tl.unfold has another order than assumed in paper
+    for iteration in range(10 ** 4):
+        A = solve_ls(tl.concatenate((tl.tenalg.khatri_rao([B, C]).T, V.T), axis=1).T,
+                     tl.concatenate((tl.unfold(X, 0), Y), axis=1).T).T
+        B = solve_ls(tl.tenalg.khatri_rao([A, C]), tl.unfold(X, 1).T).T
+        C = solve_ls(tl.tenalg.khatri_rao([A, B]), tl.unfold(X, 2).T).T
+        V = solve_ls(A, Y).T
+        f_new = calculate_f(X, Y, A, B, C, V)
```

Comment:
The code might be clearer if we remove `calculate_f` and solve inline, and use a more informative name for `f`.